File: week-4/partition-to-k-equal-sum-subsets/partition-to-k-equal-sum-subsets.py
class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        if k == 0 or sum(nums)%k != 0 or k > len(nums):
            return False
        
        return self.backtrack(0, 0, sum(nums)//k, [False]*len(nums), k, nums)
    
    def backtrack(self, index, cur_sum, total, seen, k, nums):
        if k == 1:
            return True
        
        if cur_sum == total:
            return self.backtrack(0, 0, total, seen, k-1, nums)
        
        for i in range(index, len(nums)):
            if not seen[i]:
                seen[i] = True
                if self.backtrack(i, cur_sum + nums[i], total, seen, k, nums):
                    return True
                seen[i] = False
                
        return False

File: week-4/partition-to-k-equal-sum-subsets/partition-to-k-equal-sum-subsets.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {boolean}
 */
var canPartitionKSubsets = function(nums, k) {
    const sum = nums.reduce((a,b)=>a+b, 0);
    if (k<=0 || sum%k !== 0 || k>nums.length) return false;
    
    return backtrack(0, 0, sum/k, k, nums, new Array(nums.length).fill(false));
};

const backtrack = (index, cur_sum, total, k, nums, seen) => {
    if (k === 1) return true;
    
    if (cur_sum === total) return backtrack(0, 0, total, k-1, nums, seen);
    
    for (let i=index; i<nums.length; i++) {
        if (!seen[i]) {
            seen[i] = true;
            if (backtrack(i, cur_sum + nums[i], total, k, nums, seen)) return true;
            seen[i] = false;
        }
    }
    
    return false;
}
